### ISSUE TITLE
Parallelize Path Searches

### ISSUE THREAD

**Issue by whimxiqal** [whimxiqal/journey#32]
When performing a search, Journey spins up a `SearchSession`, which creates one `ItineraryTrial` at a time, which in turn creates multiple `PathTrials`. `PathTrial`s each try to find a path between an origin and a destination and `ItineraryTrials` wait until the `PathTrials` come back before determining whether that list of paths do, in fact, form a suitable itinerary.

When trying to find an itinerary, though, the `ItineraryTrial` already knows which paths it wants. So, there's no reason each of these `PathTrials` can't be parallelized over the systems available threads. 

Changes would include:
- [ ] Adding suitable `Future` return values from a path trial's search
- [ ] Thread-safe cancel methods on each of the path-trial's search
  - [ ] If any of the paths fail during the `ItineraryTrial`, we should try to cancel all of the ongoing `PathTrial`s. If they in fact returned canceled, then we just ignore whatever else is in the result. Just like today, if any of them actually succeed even if they were cancelled, then we should still treat that like a success for caching purposes.
- [ ] Kicking off *all* of the `PathTrials` during an `ItineraryTrial` and wait for them to all complete (and cancel, if needed, as described above)
- [ ] Today, we have a config parameter that limits how many `SearchSessions` may be going at one time. Today, this effectively limits how many `PathTrials` are searching, which in turn limits how much memory the plugin can possibly use, since each of the `PathTrials` have a limited number of visited nodes during its A* search. This new functionality invalidates that assumption, since each `SearchSession` may now have multiple searches running in parallel. So, we could instead change the config parameter to something that tracks the maximum number of path algorithms running at one time (we can keep the same name, just change its description). But, we actually still have a problem: this should probably be its own ticket:
  - We can theoretically lock up all the threads on the system, blocking out all other plugins' asynchronous tasks (assuming the player's max searches config parameter is high enough). This can be fixed by instead having a maximum number of threads that can be used at a time. Since we never have multiple path trials running at once on the same thread, then this also keeps the memory usage under a maximum. So, we should do this instead.

Other notes:
  - Multi-threaded behavior is ultimately achieved by just submitting work to Bukkit (and eventually Sponge, etc.) provided async scheduler. I don't see any need to do anything custom and fancy.

**Comment by whimxiqal** [comment]
Added in #47 for v1.1
